format-specs/geoparquet.md

```diff
@@ -83,10 +81,68 @@ The optional `epoch` field allows to specify this in case the `crs` field define
 
 Note that the current version of the spec only allows for a subset of WKB: 2D or 3D geometries of the standard geometry types (the Point, LineString, Polygon, MultiPoint, MultiLineString, MultiPolygon, and GeometryCollection geometry types). This means that M values or non-linear geometry types are not yet supported.
 
+WKB geometry columns MUST be stored using the `BYTE_ARRAY` parquet type.
+
+Implementation note: when using WKB encoding with the ecosystem of Arrow libraries, Parquet types such as `BYTE_ARRAY` might not be directly accessible. Instead, the corresponding Arrow data type can be `Arrow::Type::BINARY` (for arrays that whose elements can be indexed through a 32-bit index) or `Arrow::Type::LARGE_BINARY` (64-bit index). It is recommended that GeoParquet readers are compatible with both data types, and writers preferably use `Arrow::Type::BINARY` (thus limiting to row groups with content smaller than 2 GB) for larger compatibility.
+
+##### Native encodings (based on GeoArrow)
+
+Using the single-geometry type encodings (i.e., `"point"`, `"linestring"`, `"polygon"`, `"multipoint"`, `"multilinestring"`, `"multipolygon"`) may provide better performance and enable readers to leverage more features of the Parquet format to accelerate geospatial queries (e.g., row group-level min/max statistics). These encodings correspond to extension name suffix in the [GeoArrow metadata specification for extension names](https://geoarrow.org/extension-types#extension-names) to signify the memory layout used by the geometry column. GeoParquet uses the separated (struct) representation of coordinates for single-geometry type encodings because this encoding results in useful column statistics when row groups and/or files contain related features.
+
+The actual coordinates of the geometries MUST be stored as native numbers, i.e. using
+the `DOUBLE` parquet type in a (repeated) group of fields (exact repetition depending
+on the geometry type).
+
+For the `"point"` geometry type, this results in a struct of two fields for x
+and y coordinates (in case of 2D geometries):
+
+```
+// "point" geometry column as simple field with two child fields for x and y
+optional group geometry {
+  required double x;
+  required double y;
+}
+```
+
+For the other geometry types, those x and y coordinate values MUST be embedded
+in repeated groups (`LIST` logical parquet type). For example, for the
+`"multipolygon"` geometry type:
+
+```
+// "multipolygon" geometry column with multiple levels of nesting
+optional group geometry (List) {
+  // the parts of the MultiPolygon
+  repeated group list {
+    optional group element (List) {
+      // the rings of one Polygon
+      repeated group list {
+        optional group element (List) {
+          // the list of coordinates of one ring
+          repeated group list {
+            optional group element {
+              required double x;
+              required double y;
```

Comment:
At the moment I only made the inner x and y as "required" here, but I _think_ I should also mark the other sub-levels (for polygons/rings/vertices) as required? I think we said only the geometry itself can be null in the GeoArrow spec.

Reply:
Changed all to "required" except for the top-level geometry group

Reply:
How much is the data saving? Does parquet essentially not need a nullability buffer for these arrays when the child arrays are saved as non-nullable? In my own implementation I think I'm pretty strict about having child fields marked as non-nullable, so I'm ok making this required here if you are.

Reply:
> How much is the data saving? Does parquet essentially not need a nullability buffer for these arrays when the child arrays are saved as non-nullable?

Looking at the Parquet docs, it says (https://parquet.apache.org/docs/file-format/nulls/):

> Nullity is encoded in the definition levels (which is run-length encoded). 

and then here (https://parquet.apache.org/docs/file-format/data-pages/):

> For data that is required, the definition levels are skipped

So essentially if a field is required, it can skip some data being stored (the definition levels for that field). Although given that it is run-length encoded, this data will also be tiny in case there are no nulls (because then essentially it's just one constant being run-length encoded, I assume)

See also https://arrow.apache.org/blog/2022/10/05/arrow-parquet-encoding-part-1/

Reply:
For the actual spec here, I think we can recommend `required` (and show it like that in the example) but mention that it is not required and that using `optional` fields if also fine (i.e. it's not a MUST, for example in case there is no easy way to control it depending on the Parquet implementation you're using)

Reply:
> For the actual spec here, I think we can recommend `required` (and show it like that in the example) but mention that it is not required and that using `optional` fields if also fine (i.e. it's not a MUST, for example in case there is no easy way to control it depending on the Parquet implementation you're using)

Yeah that sounds fine to me. Maybe say something that there MUST be no null values for child arrays, even if it's marked as `optional`?